NVIDIA STOCK ANALYSIS 


In [2]:
import pandas as pd
import numpy as np
import streamlit as st
import matplotlib.pyplot as plt

LOADING THE CSV FILE AND GATHERING BASIC INFO OF THE DATASET

In [4]:
st.title("Data Visualization of NVIDIA Stock with Streamlit")
st.subheader("NVIDIA Stock Price Analysis")
st.text("This application visualizes the stock price of NVIDIA over time using Streamlit.")
st.write("The data is sourced from a kaggle dataset and includes the closing prices of NVIDIA stock.")

2026-04-02 03:43:04.184 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.185 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.185 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.186 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.186 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.186 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.187 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-04-02 03:43:04.188 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bar

In [5]:
df = pd.read_csv('stocks.csv')
df.columns

FileNotFoundError: [Errno 2] No such file or directory: 'stocks.csv'

In [7]:
df.shape # (number of rows, number of columns)

(7079, 16)

In [8]:
df.dtypes

date                         object
open                        float64
high                        float64
low                         float64
close                       float64
volume                        int64
shares_outstanding_bn       float64
market_cap_usd_bn           float64
quarterly_revenue_usd_bn    float64
sma_20                      float64
sma_50                      float64
sma_200                     float64
rsi_14                      float64
stock_split                 float64
era                          object
key_event                    object
dtype: object

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7079 entries, 0 to 7078
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   date                      7079 non-null   object 
 1   open                      7079 non-null   float64
 2   high                      7079 non-null   float64
 3   low                       7079 non-null   float64
 4   close                     7079 non-null   float64
 5   volume                    7079 non-null   int64  
 6   shares_outstanding_bn     7079 non-null   float64
 7   market_cap_usd_bn         7079 non-null   float64
 8   quarterly_revenue_usd_bn  7079 non-null   float64
 9   sma_20                    7079 non-null   float64
 10  sma_50                    7079 non-null   float64
 11  sma_200                   7079 non-null   float64
 12  rsi_14                    7079 non-null   float64
 13  stock_split               6 non-null      float64
 14  era     

DATE CLEANING

In [27]:
df["date"] = pd.to_datetime(df['date'])

/var/folders/26/q8ygphr55ss1r66psnfwx_mc0000gn/T/ipykernel_3301/3435947568.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df["date"] = pd.to_datetime(df['date'])


In [28]:
df["stock_split"] = df["stock_split"].fillna(0)

In [29]:
df["key_event"] = df["key_event"].fillna("No Event")

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7079 entries, 0 to 7078
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   date                      7079 non-null   datetime64[ns]
 1   open                      7079 non-null   float64       
 2   high                      7079 non-null   float64       
 3   low                       7079 non-null   float64       
 4   close                     7079 non-null   float64       
 5   volume                    7079 non-null   int64         
 6   shares_outstanding_bn     7079 non-null   float64       
 7   market_cap_usd_bn         7079 non-null   float64       
 8   quarterly_revenue_usd_bn  7079 non-null   float64       
 9   sma_20                    7079 non-null   float64       
 10  sma_50                    7079 non-null   float64       
 11  sma_200                   7079 non-null   float64       
 12  rsi_14              

FEATURE ENGINEERING

In [26]:
df["price_change"] = df["close"] - df["open"]

df["daily_return"] = df["price_change"] / df["open"]

df["daily_return_pct"] = df["daily_return"] * 100

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7079 entries, 0 to 7078
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   date                      7079 non-null   datetime64[ns]
 1   open                      7079 non-null   float64       
 2   high                      7079 non-null   float64       
 3   low                       7079 non-null   float64       
 4   close                     7079 non-null   float64       
 5   volume                    7079 non-null   int64         
 6   shares_outstanding_bn     7079 non-null   float64       
 7   market_cap_usd_bn         7079 non-null   float64       
 8   quarterly_revenue_usd_bn  7079 non-null   float64       
 9   sma_20                    7079 non-null   float64       
 10  sma_50                    7079 non-null   float64       
 11  sma_200                   7079 non-null   float64       
 12  rsi_14              

In [30]:
positive_days = (df["price_change"] > 0).sum()
print(positive_days)

negative_days = (df["price_change"] < 0).sum()
print(negative_days)

total_days = len(df)
print(total_days)

    

3526
3484
7079


In [31]:
positive_pct = (positive_days / total_days) * 100
negative_pct = (negative_days / total_days) * 100

print(round(positive_pct, 2))
print(round(negative_pct, 2))


49.81
49.22


In [32]:
daily_avg = df["daily_return"].mean() *100
max_gain = df["daily_return"].max() * 100
min_loss = df["daily_return"].min() * 100


print(round(daily_avg, 4)) #use my 'Short/Scientific' display mode, no need to round to 6 decimal places
print(round(max_gain, 2))
print(round(min_loss, 2))

# #with f string 
# print(f"{daily_avg:.6f}")


0.0001
1.53
-1.58


STRONG BULLISH CONDITION:
If all are:
above
uptrend
golden_cross

WEAK CONDITIONS: 
If all are:
below
downtrend
no_signal


In [ ]:
#CURRENT PRICE IS GREATER THAN THE RECENT AVERAGE = ABOVE 
df["trend_signal"] = np.where(df["close"] > df["sma_20"], "Above", "Below")
df["trend_signal"].value_counts()

trend_signal
Above    3601
Below    3478
Name: count, dtype: int64

In [13]:
# RECENT PRICES ARE ABOVE THAN THE LONG TERM TREND PRICES = UPTREND 
df["trend_strength"] = np.where(df["sma_20"] > df["sma_50"], "Uptrend", "Downtrend")
df["trend_strength"].value_counts()

trend_strength
Downtrend    3599
Uptrend      3480
Name: count, dtype: int64

In [14]:
#MEDIUM TERM TREND > LONG TERM TREND = GOLDEN CROSS 
df["signal"] = np.where(df["sma_50"] > df["sma_200"], "Golden Cross", "No signal")
df["signal"].value_counts()

signal
No signal       3592
Golden Cross    3487
Name: count, dtype: int64

In [20]:
print(df[["date", "close", "trend_strength", "trend_signal", "signal"]].tail(10))

            date    close trend_strength trend_signal        signal
7069  26/02/2026  18.3312      Downtrend        Above     No signal
7070  27/02/2026  18.4086        Uptrend        Below     No signal
7071  02/03/2026  18.3975      Downtrend        Above  Golden Cross
7072  03/03/2026  18.4394        Uptrend        Below     No signal
7073  04/03/2026  18.3207      Downtrend        Above  Golden Cross
7074  05/03/2026  18.6045      Downtrend        Above     No signal
7075  06/03/2026  18.5820      Downtrend        Above  Golden Cross
7076  09/03/2026  18.5577      Downtrend        Below  Golden Cross
7077  10/03/2026  18.5336      Downtrend        Below  Golden Cross
7078  11/03/2026  18.6000        Uptrend        Below     No signal


In [21]:
print(df["trend_signal"].value_counts())
print(df["trend_strength"].value_counts())
print(df["signal"].value_counts())

trend_signal
Above    3601
Below    3478
Name: count, dtype: int64
trend_strength
Downtrend    3599
Uptrend      3480
Name: count, dtype: int64
signal
No signal       3592
Golden Cross    3487
Name: count, dtype: int64


Strong bullish condition across all timeframes


In [22]:
condition = (
    (df["trend_signal"] == "Above") &
    (df["trend_strength"] == "Uptrend") &
    (df["signal"] == "Golden Cross")
)


In [23]:
strong_bullish = condition.sum()
print(strong_bullish)


472


In [24]:
strong_bullish_pct = (strong_bullish / len(df)) * 100
print(round(strong_bullish_pct, 2))

6.67


VOLATILITY - how much the price fluctuates over time

In [56]:
volatility = np.std(df["daily_return"].to_numpy())
volatility_pct = volatility * 100

print(f"Volatility: {volatility_pct:.3f}%")

Volatility: 0.443%


On average, NVIDIA stock fluctuates by about 0.443% per day.

In [67]:
uptrend_mask = (df["trend_strength"] == "Uptrend")
downtrend_mask = (df["trend_strength"] == "Downtrend")

#filtering the daily returns based on the trend strength
uptrend_returns = df["daily_return"][uptrend_mask]
downtrend_returns = df["daily_return"][downtrend_mask]

#calculating volatility for uptrends and downtrends
up_vol = np.std(uptrend_returns) * 100
down_vol = np.std(downtrend_returns) * 100

#displaying the volatility for uptrends and downtrends
print(f"Uptrend Volatility: {up_vol:.3f}%")
print(f"Downtrend Volatility: {down_vol:.3f}%")

Uptrend Volatility: 0.446%
Downtrend Volatility: 0.441%
